### PR TITLE
Physical Plan for OFFSET

### PR DESCRIPTION
**Pull request by ming535**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/2551

 # Rationale for this change

Support physical plan for `OFFSET`.
Tow test cases for `OFFSET` without `LIMIT` are still not working, which is related to `limit_push_down` https://github.com/apache/arrow-datafusion/issues/2624

# What changes are included in this PR?

Implement physical plan of `OFFSET`

# Are there any user-facing changes?

# Does this PR break compatibility with Ballista?

### CODE REVIEW

File: datafusion/core/tests/sql/offset.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use super::*;
+
+#[tokio::test]
+async fn csv_offset_without_limit() -> Result<()> {
```
Comment (ming535) [inline datafusion/core/tests/sql/offset.rs:21]: not working right now.

File: datafusion/core/src/optimizer/limit_push_down.rs [hunk 2]
```diff
@@ -386,6 +386,19 @@ mod test {
         Ok(())
     }
 
+    #[test]
+    fn limit_pushdown_should_not_pushdown_limit_with_offset_only() -> Result<()> {
```
Comment (ming535) [inline datafusion/core/src/optimizer/limit_push_down.rs:390]: This test case is not working right now due to limit_push_down

**Comment by ming535** [comment]
I think a better way to solve this issue https://github.com/apache/arrow-datafusion/issues/2624 might be to add another LogicalPlan, for example `LimitWithOffset`. The semantic of these three LogicalPlan is different:

1. `Limit`: only set an upper bounds on the stream's data
2. `Offset` only set a lower bounds on the stream's data
3. `LimitWithOffset` set a lower bounds AND upper bounds on the stream's data

As of `limit_push_down`, it wont' touch `Offset`.
It will push down `limit` of `Limit`.
It will adjust `LimitWithOffset`'s  `limit` based on its offset and push the limit down to TableScan; if the push down happens, we replace it with `Offset`. 

As of physical plan, we need to implement 3 physical plan accordingly. (Or may be we should just )

I am not quite sure whether this is a good way to solve this issue, any thoughts?
@alamb @andygrove @Ted-Jiang

File: datafusion/core/src/physical_plan/offset.rs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+                _ => {
+                    // continue to poll input stream
+                }
+            }
+        }
+    }
+}
+
+/// Remove a RecordBatch's first n rows
+pub fn cut_batch(batch: &RecordBatch, n: usize) -> RecordBatch {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/offset.rs:242]: I think you might be able to use `RecordBatch::slice()` here

https://docs.rs/arrow/14.0.0/arrow/record_batch/struct.RecordBatch.html#method.slice

**Review by alamb — COMMENTED** [review]
Thanks @ming535  -- I skimmed this PR briefly.

> I think a better way to solve this issue https://github.com/apache/arrow-datafusion/issues/2624 might be to add another LogicalPlan, for example LimitWithOffset. The semantic of these three LogicalPlan is different:

I agree the semantics are different,

What about using a single physical operator for all three cases : Limit, Offset, and LimitWithOffset? 

The difference in the actual logic is quite small and we could avoid the 100s of lines of code for a new physical operator.

**Comment by ming535** [comment]
Will update the PR when https://github.com/apache/arrow-datafusion/pull/2638 is reviewed and merged.

**Comment by ming535** [comment]
If https://github.com/apache/arrow-datafusion/pull/2694 is ok, then this PR can be closed.

**Pull request closed without merging**